### PR TITLE
[uma] allow building UMA as shared library

### PR DESCRIPTION
**Pull request by igchor**
This might be useful for prototype integration with OpenMP

### CODE REVIEW

File: source/common/unified_memory_allocation/CMakeLists.txt [hunk 1]
```diff
@@ -1,11 +1,19 @@
...
 # SPDX-License-Identifier: MIT
 
-add_library(unified_memory_allocation STATIC
+set(UMA_SOURCES
     src/memory_pool.c
     src/memory_provider.c
 )
 
+if(UMA_BUILD_SHARED_LIBRARY)
+    add_library(unified_memory_allocation SHARED
```
Comment (pbalcer) [inline source/common/unified_memory_allocation/CMakeLists.txt:10]: `message(WARNING, "Unified Memory Allocation is still an early work in progress. There are no API/ABI backward compatibility guarantees. There will be breakages. Do not use the shared library in production software.")`

Or something to that effect.

File: source/common/unified_memory_allocation/CMakeLists.txt [hunk 2]
```diff
@@ -1,11 +1,19 @@
 # Copyright (C) 2023 Intel Corporation
 # SPDX-License-Identifier: MIT
 
-add_library(unified_memory_allocation STATIC
+set(UMA_SOURCES
     src/memory_pool.c
     src/memory_provider.c
 )
 
+if(UMA_BUILD_SHARED_LIBRARY)
```
Comment (kbenzie) [inline source/common/unified_memory_allocation/CMakeLists.txt:9]: Is `USM_BUILD_SHARED_LIBRARY` only defined when the user sets it during CMake configuration?

Making it an [option](https://cmake.org/cmake/help/latest/command/option.html) would be nice.

File: source/common/unified_memory_allocation/CMakeLists.txt (hunk 1 above)
Comment (igchor) [reply source/common/unified_memory_allocation/CMakeLists.txt:13]: Done.

File: source/common/unified_memory_allocation/CMakeLists.txt (hunk 2 above)
Comment (igchor) [reply source/common/unified_memory_allocation/CMakeLists.txt:9]: Right, done.

**Review by pbalcer — APPROVED** [review]

**Review by vinser52 — APPROVED** [review]

**Pull request merged**